internal/providers/github/factory/factory.go

```diff
@@ -0,0 +1,223 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package factory contains the GitHubProviderFactory
+package factory
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"slices"
+
+	gogithub "github.com/google/go-github/v61/github"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+
+	"github.com/stacklok/minder/internal/config/server"
+	"github.com/stacklok/minder/internal/crypto"
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/providers/credentials"
+	"github.com/stacklok/minder/internal/providers/factory"
+	githubapp "github.com/stacklok/minder/internal/providers/github/app"
+	ghclient "github.com/stacklok/minder/internal/providers/github/oauth"
+	"github.com/stacklok/minder/internal/providers/ratecache"
+	"github.com/stacklok/minder/internal/providers/telemetry"
+	v1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+// NewGitHubProviderClassFactory creates an instance of ProviderClassFactory
+// for creating GitHub-specific provider instances.
+func NewGitHubProviderClassFactory(
+	restClientCache ratecache.RestClientCache,
+	metrics telemetry.HttpClientMetrics,
+	config *server.GitHubAppConfig,
+	fallbackTokenClient *gogithub.Client,
+	crypteng crypto.Engine,
+	store db.Store,
+) factory.ProviderClassFactory {
+	return &githubProviderFactory{
+		restClientCache:     restClientCache,
+		metrics:             metrics,
+		config:              config,
+		fallbackTokenClient: fallbackTokenClient,
+		crypteng:            crypteng,
+		store:               store,
+	}
+}
+
+type githubProviderFactory struct {
+	restClientCache     ratecache.RestClientCache
+	metrics             telemetry.HttpClientMetrics
+	config              *server.GitHubAppConfig
+	fallbackTokenClient *gogithub.Client
+	crypteng            crypto.Engine
+	store               db.Store
+}
+
+var (
+	supportedClasses = []db.ProviderClass{
+		db.ProviderClassGithubApp,
+		db.ProviderClassGithub,
+	}
+)
+
+func (_ *githubProviderFactory) GetSupportedClasses() []db.ProviderClass {
+	return supportedClasses
+}
+
+func (g *githubProviderFactory) Build(ctx context.Context, config *db.Provider) (v1.Provider, error) {
+	class := config.Class
+	// This should be validated by the caller, but let's check anyway
+	if !slices.Contains(supportedClasses, class) {
+		return nil, fmt.Errorf("provider does not implement github")
+	}
+
+	if config.Version != v1.V1 {
+		return nil, fmt.Errorf("provider version not supported")
+	}
+
+	creds, err := g.getProviderCredentials(ctx, config)
```

Comment:
There are some places where we don't use the DB credential and pass a different credential into the provider constructor.
I think it's valid to not use a provider in those cases, or pass the credential into the function call. 
I'm only calling it out here to be aware of it when we refactor that piece of code later.